# Intelligent Systems SIN 2025/26, lab group 3E1, subgroup **2**.
# **A3**. Block 2 lab exam (1.25 points).
## 17 December 2025. **TIME: 45 min**.

### Student:
Adrià Marín Bayarri

In [1]:
pip install scikit-learn


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


Before starting this exam, you must download the dataset (samples **X.npy** and labels **y.npy**) and the logistic regression library (**Logistic_Regression.py**) from Poliformat.

In [ ]:
# You only need to execute this cell if running in Google Colab 

# You need to upload LogisticRegression.py
from google.colab import files
uploaded = files.upload()

# You need to upload X.npy 
from google.colab import files
uploaded = files.upload()

# You need to upload y.npy
from google.colab import files
uploaded = files.upload()

**Dataset description**: This dataset contains **1500** samples with **254** features, for a classification task into **29** classes.

By running the code below, you will load the dataset and partition it into 80% of the samples for training and 20% for testing (with seed random_state$=$23), and you will obtain a baseline test error rate (of 12.3%) from a logistic regression classifier trained using default parameters and batch size 10.

In [2]:
from LogisticRegression import LogisticRegressionClassification, LogisticRegressionTraining
import numpy as np
from sklearn.model_selection import train_test_split

# Load data
X = np.load('X.npy'); y = np.load('y.npy')
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=True, random_state=23)
N = len(X_train); M = len(X_test)

# Train and classify
W = LogisticRegressionTraining(X_train, y_train, bs=10)
haty_test = LogisticRegressionClassification(X_test, W)
accuracy = np.sum(haty_test==y_test)/M
print(f"Test error: {1.0-accuracy:.1%}")

Test error: 12.3%


### Exercise 1
Applying the logistic regression classifier with default parameter values (eta 1e-2, batch size 10), explore the maximum number of epochs in logarithmic scale to determine an optimal value.

According to the results:

1.1) What is the optimal value for the maximum numer of epochs? How do you decide which one to choose?

1.2) Is overfitting/overtraining observed? If so, from what epoch?

In [3]:
print("  bs maxEps     eta trainErr testErr")
print("---- ------ ------- -------- -------")
bs=10; eta=1e-2;
for maxEpochs in (5, 10, 20, 50, 100, 200, 500, 1000, 2000):
    W = LogisticRegressionTraining(X_train,y_train, bs=bs, maxEpochs=maxEpochs, eta=eta)
    haty_train = LogisticRegressionClassification(X_train,W)
    acc_train = np.sum(haty_train==y_train)/N
    haty_test = LogisticRegressionClassification(X_test,W)
    acc_test = np.sum(haty_test==y_test)/M
    print(f"{bs:4d} {maxEpochs:6d} {eta:.1e} {1.0-acc_train:8.1%} {1.0-acc_test:7.1%}")

  bs maxEps     eta trainErr testErr
---- ------ ------- -------- -------
  bs maxEps     eta trainErr testErr
---- ------ ------- -------- -------
  10      5 1.0e-02     5.2%   12.3%
  10     10 1.0e-02     4.0%   12.3%
  10     20 1.0e-02     2.0%    9.3%
  10     50 1.0e-02     0.7%   11.3%
  10    100 1.0e-02     0.2%   12.3%
  10    200 1.0e-02     0.0%   12.3%
  10    500 1.0e-02     0.0%   12.0%
  10   1000 1.0e-02     0.0%   12.0%
  10   2000 1.0e-02     0.0%   12.0%


**Answers:**

1.1) For the best maxEpochs we will choose the one value that reduces both the trainErr and testErr to the maximum possible, without overfitting the model. In our case, we can observe that the testError decreases until maxEpochs = 50, where it increases again, which could mean that the model is specializing too much in classifying our training data and loosing generalisation. For this reason, I would choose maxEpochs = 20, as it accomplishes a good balance between low trainError and testError.

1.2) Yes, there is overfitting from Epoch = 50 forward.

### Exercise 2
Using maximum number of epochs 1000, batch size (bs) 10 and applying the logistic regression classifier, adjust the learning rate (eta) in logarithmic scale to determine an optimal value. Report the classification error rate on the training and test sets.

Based on the results, what is the optimal value for the learning rate? Discuss the results obtained.

In [4]:
print("  bs maxEps     eta trainErr testErr")
print("---- ------ ------- -------- -------")

bs=10; maxEpochs = 1000;
for eta in (1e-4, 1e-3, 1e-2, 1e-1, 1, 1e1, 1e2):
    W = LogisticRegressionTraining(X_train,y_train, bs=bs, maxEpochs=maxEpochs, eta=eta)
    haty_train = LogisticRegressionClassification(X_train,W)
    acc_train = np.sum(haty_train==y_train)/N
    haty_test = LogisticRegressionClassification(X_test,W)
    acc_test = np.sum(haty_test==y_test)/M
    print(f"{bs:4d} {maxEpochs:6d} {eta:.1e} {1.0-acc_train:8.1%} {1.0-acc_test:7.1%}")

  bs maxEps     eta trainErr testErr
---- ------ ------- -------- -------
  10   1000 1.0e-04     3.1%    9.3%
  10   1000 1.0e-03     0.2%   12.0%
  10   1000 1.0e-02     0.0%   12.0%
  10   1000 1.0e-01     2.7%   14.7%
  10   1000 1.0e+00     9.8%   22.0%
  10   1000 1.0e+01    10.4%   15.0%
  10   1000 1.0e+02     8.3%   16.0%


**Answer:**

2.1) The learning rate adjusts the speed at which our model updates the weights to the optimal or correct value for prediction. The higher the learning rate, the greater will be the "jumps" it will perform. However, using high values of learning rate can make that a model changes too much, leading to incorrect results. For this reason, we should find a learning rate that correctly converges to one set of weights, decreasing the errors in the prediction. In this case, the learning rate (eta) that best fits to this description might be eta=1.0e-04, as it provides the lowest testError from all the learning rates. Its trainError, although it is not 0%, it is sufficiently low to provide us good results.  

### Exercise 3
Applying the logistic regression classifier with default parameter values and
learning rate (eta) 1e-2, explore at the same time both the maximum number of epochs and the batch size
(bs) to determine their optimal values together (tip: use a nested loop, with bs in the outer loop for easier interpretability). Use in both cases a logarithmic
scale. Report the classification error rate on the training and test
sets.

Based on the results, what are the optimal values? Discuss the results obtained.

In [5]:
print("  bs maxEps     eta trainErr testErr")
print("---- ------ ------- -------- -------")
eta=1e-2;

for bs in (1, 2, 5, 10, 20, 50, 100):
    for maxEpochs in (5, 10, 20, 50, 100, 200, 500, 1000, 2000):
        W = LogisticRegressionTraining(X_train,y_train, bs=bs, maxEpochs=maxEpochs, eta=eta)
        haty_train = LogisticRegressionClassification(X_train,W)
        acc_train = np.sum(haty_train==y_train)/N
        haty_test = LogisticRegressionClassification(X_test,W)
        acc_test = np.sum(haty_test==y_test)/M
        print(f"{bs:4d} {maxEpochs:6d} {eta:.1e} {1.0-acc_train:8.1%} {1.0-acc_test:7.1%}")

  bs maxEps     eta trainErr testErr
---- ------ ------- -------- -------
   1      5 1.0e-02    21.8%   25.7%
   1     10 1.0e-02    21.8%   25.7%
   1     20 1.0e-02    21.8%   25.7%
   1     50 1.0e-02    21.8%   25.7%
   1    100 1.0e-02    21.8%   25.7%
   1    200 1.0e-02    21.8%   25.7%
   1    500 1.0e-02    21.8%   25.7%
   1   1000 1.0e-02    21.8%   25.7%
   1   2000 1.0e-02    21.8%   25.7%
   2      5 1.0e-02    10.7%   20.7%
   2     10 1.0e-02     7.4%   16.0%
   2     20 1.0e-02     1.5%   13.3%
   2     50 1.0e-02     1.5%   13.3%
   2    100 1.0e-02     1.5%   13.3%
   2    200 1.0e-02     1.5%   13.3%
   2    500 1.0e-02     1.5%   13.3%
   2   1000 1.0e-02     1.5%   13.3%
   2   2000 1.0e-02     1.5%   13.3%
   5      5 1.0e-02     5.8%   14.7%
   5     10 1.0e-02     3.9%   12.0%
   5     20 1.0e-02     1.3%   10.3%
   5     50 1.0e-02     0.2%   12.7%
   5    100 1.0e-02     0.0%   12.7%
   5    200 1.0e-02     0.0%   12.0%
   5    500 1.0e-02     0.0%   12.0%
 

**Answer:**

There are multiple combinations that offer good results for the same learning rate: 
    -bs = 10, maxEpochs = 20, trainErr = 2.0%, testErr: 9.3%
    -bs = 20, maxEpochs = 50, trainErr = 1.3%, testErr: 10%
    -bs = 20, maxEpochs = 20, trainErr = 2.9%, testErr: 9.3%
Nevertheless, the option I would choose as the most optimal would be bs=10 and maxEpochs=20, as it is the combination that accomplishes the best  testError(9.3%) without sacrificing the trainingError(2.0%), obtaining a good balance between correct predictions and generalisation(no overfitting of the model). 

### Exercise 4
According to the results you have obtained, can you claim whether this task is linearly separable? Which result(s) can you look at for that?

**Answer:**

Yes, we can state that the classes are linearly separable, as we can observe in some combinations of bs and maxEpochs a trainingError = 0.0%. Some examples of this are combinations:
    -bs = 50, maxEpochs = 1000
    -bs = 100, maxEpochs = 2000
    -bs = 5, maxEpochs = 100

If there wasn't any combination achieving this 0% in the trainingError, we wouldn't be able to determine if there is linear separation between classes, although it would still be possible